### PR TITLE
test(qemu): replace Fedora Cloud image with Alpine Linux tiny image

### PR DESCRIPTION
**Pull request by mangelajo**
## Motivation

Fedora Cloud qcow2 images used in QEMU driver tests weigh ~556 MB (x86_64) + ~519 MB (aarch64) = **~1.1 GB total**, taking ~1 minute to download on every CI run. Alpine Linux 3.22.4 UEFI tiny images reduce that to **~278 MB total** — a 74% reduction — and download in ~6 s on GitHub Actions runners, making a cache unnecessary.

| | Fedora 43 | Alpine 3.22.4 tiny | Reduction |
|---|---|---|---|
| x86\_64 | 556 MB | 127 MB | 77% |
| aarch64 | 519 MB | 151 MB | 71% |
| **Total** | **~1,076 MB** | **~278 MB** | **74%** |

Alpine's `nocloud` tiny images support UEFI boot on both architectures and use `tiny-cloud` (a minimal cloud-init alternative) that reads the standard NoCloud CIDATA vfat volume the QEMU driver already generates.

## Driver changes (`driver.py`)

`tiny-cloud` differs from full cloud-init in two ways:

1. **Hostname**: tiny-cloud reads `@hostname` from the YAML key `hostname` in `meta-data`, not `local-hostname`. Add `hostname` alongside `local-hostname` so both implementations set it correctly.

2. **Password**: tiny-cloud ignores `plain_text_passwd` in the `users` stanza. Add a `runcmd` entry to set the password via `chpasswd`. Credentials are escaped with `shlex.quote()` so special characters are safe. Full cloud-init images are unaffected.

**`hostport=0` support**: `Hostfwd` now allows `hostport=0`, which tells QEMU's user-mode networking to pick a free port automatically. After QEMU starts, the driver queries the actual assigned port via QMP `human-monitor-command info usernet`, logs it at INFO level:

```
hostfwd 'ssh': resolved port 0 -> 127.0.0.1:<port> (guest port 22)
```

and stores it in `_resolved_hostports` so `get_hostfwd_port()` can return it to the client side.

## Client changes (`client.py`)

`shell()` now checks for an `ssh` hostfwd entry by calling `get_hostfwd_port('ssh')`. If present, it opens a direct `fabric.Connection` to `127.0.0.1:<resolved_port>`, bypassing the jumpstarter streaming layer entirely. If no ssh hostfwd is configured (`KeyError`), it falls back to the original `FabricAdapter` path over vsock.

This is necessary because **Alpine uses OpenRC, not systemd**, so `systemd-ssh-generator` never runs and sshd only listens on TCP. Fedora cloud images ship `systemd-ssh-generator` (systemd ≥ 256) which automatically binds sshd to `AF_VSOCK` port 22 inside VMs — that is why the original vsock path worked with Fedora but not Alpine.

## Test changes (`driver_test.py`)

- Pass `hostfwd={'ssh': {'hostport': 0, 'guestport': 22}}` so `qemu.shell()` uses the TCP path with a QEMU-assigned port; no probe socket needed.
- Wait for `bootstrap_complete: done` on the serial console before attempting login, guaranteeing tiny-cloud has set the password and sshd is ready (avoids the race that existed with Fedora too, where the extra `setenforce` console round-trip happened to provide enough delay).
- Press Enter if the GRUB countdown appears to skip the 10 s wait.
- Update post-login prompt from bash `[user@host ~]$` → Alpine ash `host:~$`.
- Remove `sudo setenforce 0` (no SELinux on Alpine).
- Relax `uname -r` assertion to a non-empty check.

## CI changes (`python-tests.yaml`)

Remove the `actions/cache` steps entirely. Alpine images download in ~6 s on GitHub runners (measured from CI logs), which is faster than cache restore overhead for files this size.

### CODE REVIEW

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py [hunk 1]
```diff
@@ -528,6 +529,13 @@ def cidata(self) -> TemporaryDirectory:
                             "sudo": "ALL=(ALL) NOPASSWD:ALL",
                         }
                     ],
+                    # runcmd sets the password explicitly for cloud-init implementations
+                    # that do not support plain_text_passwd (e.g. Alpine's tiny-cloud).
+                    # cloud-init ignores runcmd entries it doesn't understand, so this
+                    # is safe to include unconditionally.
+                    "runcmd": [
+                        f"echo '{self.username}:{self.password}' | chpasswd",
+                    ],
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py:562]: done

**Review by raballew — CHANGES_REQUESTED** [review]
On a high level, I don't like hardcoding this in so many different places plus we are completely missing out the renovate process here. I think we will need to ensure that this image gets constantly updated as well.

**Comment by mangelajo** [comment]
> On a high level, I don't like hardcoding this in so many different places plus we are completely missing out the renovate process here. I think we will need to ensure that this image gets constantly updated as well.

no real need for this, this does not go into production or anything, is just an image that we boot to run the tests. But that's an existing problem from the past, nothing new.

File: .github/workflows/python-tests.yaml [hunk 2]
```diff
@@ -120,19 +120,11 @@ jobs:
         run: |
           brew install renode/tap/renode
 
-      - name: Cache Fedora Cloud images
```
Comment (mangelajo) [inline .github/workflows/python-tests.yaml:123]: the cache was actually slower than the download.

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/client.py [hunk 3]
```diff
@@ -75,12 +76,25 @@ def novnc(self):
...
-        with FabricAdapter(
-            client=self.ssh,
-            user=self.username,
-            connect_kwargs={"password": self.password},
-        ) as conn:
-            yield conn
+        # If the driver has an 'ssh' hostfwd entry, fetch the actual host port
+        # (resolving any port=0 assignment) and connect directly over TCP.
+        # Otherwise fall back to tunnelling through the jumpstarter stream (vsock).
+        try:
```
Comment (mangelajo) [inline python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/client.py:82]: we had to implement this because alpine does not setup ssh over vhost, and we didn't want to pick a port that could potentially conflict with whatever is on the testing host.

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/client.py [hunk 4]
```diff
@@ -75,12 +76,25 @@ def novnc(self):
...
+            port = int(self.call("get_hostfwd_port", "ssh"))
+            with Connection(
+                host="127.0.0.1",
+                port=port,
+                user=self.username,
+                connect_kwargs={"password": self.password},
+            ) as conn:
+                yield conn
+        except KeyError:
```
Comment (raballew) [inline python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/client.py:91]: Okay this one is a little backwards but hear me out: I think `shell()` method catches `KeyError` to fall back from tcp to vsock, but this never actually works. On the driver side, `get_hostfwd_port()` raises `KeyError`, but the gRPC transport catches all exceptions and reraises them as `DriverError` via `StatusCode.UNKNOWN` but `DriverError` doesn't inherit from `KeyError` so VMs without SSH hostfwd will crash with an unhandled `DriverError` instead of falling back to vsock.

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py [hunk 5]
```diff
@@ -381,6 +382,26 @@ async def on(self) -> None:  # noqa: C901
...
+                    for k, v in zero_fwds.items():
+                        if v.hostaddr == actual_hostaddr and v.guestport == actual_guestport:
+                            self.logger.info(
+                                "hostfwd '%s': resolved port 0 -> %s:%d (guest port %d)",
+                                k, actual_hostaddr, actual_hostport, actual_guestport,
+                            )
+                            self.parent._resolved_hostports[k] = actual_hostport
+
         await qmp.execute("system_reset")
         await qmp.disconnect()
```
Comment (raballew) [inline python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py:407]: A `ValueError` from malformed output or a QMP error leaves the connection open and the QEMU process orphaned. You could wrap the QMP operation block in try/finally to guarantee `await qmp.disconnect()` runs.

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py [hunk 6]
```diff
@@ -410,7 +431,7 @@ def close(self):
 class Hostfwd(BaseModel):
     protocol: Literal["tcp"] = "tcp"
     hostaddr: str = "127.0.0.1"
-    hostport: int = Field(ge=1, le=65535)
+    hostport: int = Field(ge=0, le=65535)  # 0 = let QEMU pick a free port
     guestport: int = Field(ge=1, le=65535)
```
Comment (raballew) [inline python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py:437]: If `on()` fails after process creation but before usernet resolution completes, `get_hostfwd_port()` returns a stale port. I think adding `self.parent._resolved_hostports.clear()` at the beginning of `on()` before starting the new QEMU process, and also in `off()` alongside the existing cleanup should resolve this.

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py [hunk 7]
```diff
@@ -410,7 +431,7 @@ def close(self):
 class Hostfwd(BaseModel):
     protocol: Literal["tcp"] = "tcp"
     hostaddr: str = "127.0.0.1"
-    hostport: int = Field(ge=1, le=65535)
+    hostport: int = Field(ge=0, le=65535)  # 0 = let QEMU pick a free port
```
Comment (raballew) [inline python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py:434]: [HIGH] `hostaddr: str = "127.0.0.1"` has no format validation. The value is interpolated unsanitized into a comma-separated QEMU `-netdev` argument. Since QEMU uses commas to delimit netdev options, a crafted `hostaddr` like `127.0.0.1,hostfwd=tcp:0.0.0.0:4444-:4444` would inject an additional hostfwd rule. The `hostport` and `guestport` fields have `Field(ge=..., le=...)` constraints, but `hostaddr` accepts arbitrary strings.

Suggested fix: add a Pydantic `field_validator` using `ipaddress.ip_address()` to ensure `hostaddr` is a valid IP address.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py [hunk 8]
```diff
@@ -528,12 +552,30 @@ def cidata(self) -> TemporaryDirectory:
...
 
+    @export
+    @validate_call(validate_return=True)
+    def get_hostfwd_port(self, key: str) -> int:
+        """Return the actual host port for a hostfwd entry (resolves port 0 assignments)."""
+        if key in self._resolved_hostports:
+            return self._resolved_hostports[key]
+        if key in self.hostfwd:
+            return self.hostfwd[key].hostport
+        raise KeyError(f"hostfwd key {key!r} not found")
```
Comment (raballew) [inline python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py:577]: [HIGH] `get_hostfwd_port()` is a new exported method with three code paths (resolved port, configured port, KeyError), and the `info usernet` parsing logic uses positional string splitting with `int(parts[3])` and `int(parts[5])`. Neither has any unit test. The only coverage is the full integration test that boots a real VM. Extracting the parsing logic into a standalone function (e.g., `_parse_usernet_hostfwd(usernet_output, zero_fwds)`) would make it straightforward to add parameterized unit tests covering all three `get_hostfwd_port()` code paths as well as edge cases in the usernet output format.

_AI-generated, human reviewed_

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py [hunk 9]
```diff
@@ -440,6 +461,8 @@ class Qemu(Driver):
     flash_timeout: int = field(default=30 * 60)  # 30 minutes
 
     _tmp_dir: TemporaryDirectory = field(init=False, default_factory=TemporaryDirectory)
+    # Maps hostfwd key -> actual host port after QEMU resolves port 0 assignments
+    _resolved_hostports: dict[str, int] = field(init=False, default_factory=dict)
 
     @classmethod
     def client(cls) -> str:
```
Comment (raballew) [inline python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py:468]: Defer TcpNetwork child creation for hostport=0 entries until after port resolution in `on()`, or update the child's port after resolution. Use a different key namespace for hostfwd children to avoid overwriting vsock children.

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py [hunk 10]
```diff
@@ -381,6 +382,26 @@ async def on(self) -> None:  # noqa: C901
...
+                if len(parts) >= 6 and "HOST_FORWARD" in parts[0]:
+                    # parts: Protocol[State] fd hostaddr hostport guestaddr guestport ...
+                    actual_hostaddr, actual_hostport, actual_guestport = parts[2], int(parts[3]), int(parts[5])
+                    for k, v in zero_fwds.items():
+                        if v.hostaddr == actual_hostaddr and v.guestport == actual_guestport:
+                            self.logger.info(
+                                "hostfwd '%s': resolved port 0 -> %s:%d (guest port %d)",
+                                k, actual_hostaddr, actual_hostport, actual_guestport,
+                            )
+                            self.parent._resolved_hostports[k] = actual_hostport
```
Comment (raballew) [inline python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver.py:403]: The `info usernet` parser uses `int(parts[3])` and `int(parts[5])` without try/except. A malformed line would raise `ValueError

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/client.py [hunk 11]
```diff
@@ -4,6 +4,7 @@
 from contextlib import contextmanager
 
 import click
+from fabric import Connection
```
Comment (raballew) [inline python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/client.py:7]: `from fabric import Connection` is a new import added by this PR, but `fabric` is not listed in dependencies.

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/client.py [hunk 12]
```diff
@@ -75,12 +76,25 @@ def novnc(self):
...
+                connect_kwargs={"password": self.password},
+            ) as conn:
+                yield conn
+        except KeyError:
+            with FabricAdapter(
+                client=self.ssh,
+                user=self.username,
+                connect_kwargs={"password": self.password},
+            ) as conn:
+                yield conn
```
Comment (raballew) [inline python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/client.py:97]: Narrow the try/except to only wrap the `self.call("get_hostfwd_port", "ssh")` call, then branch on the result.

File: python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver_test.py [hunk 13]
```diff
@@ -88,16 +93,22 @@ def test_driver_qemu(tmp_path, ovmf):
...
             p.expect_exact("Password:")
             p.sendline(password)
-            p.expect_exact(f"[{username}@{hostname} ~]$")
-            p.sendline("sudo setenforce 0")
-            p.expect_exact(f"[{username}@{hostname} ~]$")
+            p.expect_exact(f"{hostname}:~$")
 
         with qemu.shell() as s:
-            assert s.run("uname -r").stdout.strip() == f"6.17.1-300.fc43.{arch}"
+            assert s.run("uname -r").stdout.strip() != ""
```
Comment (raballew) [inline python/packages/jumpstarter-driver-qemu/jumpstarter_driver_qemu/driver_test.py:111]: Also assert the return code, e.g., `result = s.run("uname -r"); assert result.ok; assert result.stdout.strip() != ""`.

**Comment by mangelajo** [comment]
I need to review the qemu/ssh changes more carefully

**Review by raballew — CHANGES_REQUESTED** [review]

**Comment by mangelajo** [comment]
will close for now
